python/lsst/meas/extensions/shapeHSM/_hsm_moments.py

```diff
@@ -0,0 +1,208 @@
+class HsmSourceMomentsPlugin(HsmMomentsPlugin):
+    ConfigClass = HsmSourceMomentsConfig
+
+    def __init__(self, config, name, schema, metadata, logName=None) -> None:
+        super().__init__(config, name, schema, logName)
+        # some flag thingy
+        # work in conjunction with fail()
+        self.centroidResultKey = afwTable.Point2DKey.addFields(
+            schema, "centroid", "The centroid measured by HSM", "pixel"
+        )
+        self.shapeKey = afwTable.QuadrupoleKey.addFields(schema, name, doc="HSM source moments.")
+        if config.addFlux:  # search the correct attribute name in the codebase
+            self.fluxKey = measBase.FluxResultKey.addFields(schema, "flux", "The flux measured by HSM", "dn")
+
+    def measure(self, record, exposure):
+        """... in place"""
+        # source in c++ is record in py. exposure the same
+
+        center = record.getCentroid()
+        badMaskPlanes = self.config.badMaskPlanes
+        badPixelMask = exposure.getMaskedImage().getMask().getPlaneBitMask(badMaskPlanes)
+        bbox = record.getFootprint().getBBox()
+        if bbox.getArea() == 0:
+            raise MeasurementError("Find the same error message from .cc file")
+
+        if not bbox.contains(center):
+            raise MeasurementError("Find the same error message from .cc file")
+
+        psfSigma = exposure.getPsf().computeShape(center).getTraceRadius()
+
+        # self._calculate(record, exposure, bbox, center, badPixelMask, ...)
+
+        # galsim
+        image_array = exposure[bbox].getImage().array
+        xmin, xmax = bbox.getMinX() + 1, bbox.getMaxX() + 1
+        ymin, ymax = bbox.getMinY() + 1, bbox.getMaxY() + 1
+        bounds = galsim.bounds.BoundsI(xmin, xmax, ymin, ymax)
+        image = galsim.Image(image_array, bounds=bounds, copy=False)
+        # GalSim's HSM uses the FITS convention of 1,1 for the lower-left corner
+        # account for difference in origin while converting center to galsim centerid (guessed)
+        # make sure bbox is centered properly
+
+        try:
+            shape = galsim.hsm.FindAdaptiveMom(
+                image,
+                weight=None,
+                badpix=badPixelMask,
+                guess_sig=2.5 * psfSigma,
+                precision=1.0e-6,
+                guess_centroid=None,
+                strict=True,
+                round_moments=False,
+                hsmparams=None,
+                use_sky_coords=False,
+            )
+
+        except galsim.hsm.GalSimHSMError as e:
+            raise MeasurementError(e)
+
+        # now that we have shape, we can get the values below:
+        determinantRadius = shape.moments_sigma
+        centroidResult = shape.moments_centroid
+
+        # subtract center should be implemented as well
+        if self.config.subtractCenter:
+            centroidResult.x -= center.getX()
+            centroidResult.y -= center.getY()
+
+        record.set(self.centroidResultKey, centroidResult)
```

Comment:
`centroidResult` would be an instance of a `galsim` class, and should ideally be a `lsst.geom.Point2D`. I'm surprised this didn't result in an error.